### PR TITLE
sql: add FunctionReferenceResolver interface

### PR DESCRIPTION
**Pull request by mgartner**
A `FunctionReferenceResolver` interface has been added that provides the
ability to resolve a built-in or user-defined function definition from
an unresolved name. This interface is currently unused, but will be used
in future commits to implement UDF resolution. It will also allow us to
write optimizer tests for UDFs with the test catalog. We can implement a
test-only `FunctionReferenceResolver` to resolve functions defined in
`exec-ddl` commands.

Release note: None

### CODE REVIEW

File: pkg/sql/sem/tree/type_check.go [hunk 1]
```diff
@@ -1036,7 +1040,11 @@ func (expr *FuncExpr) TypeCheck(
 	if semaCtx != nil {
 		searchPath = semaCtx.SearchPath
 	}
-	def, err := expr.Func.Resolve(searchPath)
+	var resolver FunctionReferenceResolver
+	if semaCtx != nil {
```
Comment (chengxiong-ruan) [inline pkg/sql/sem/tree/type_check.go:1044]: nit: merge this `if` block with the one setting search path above?

**Review by chengxiong-ruan — APPROVED** [review]
LGTM, thanks for doing this.

**Review by msirek — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 22 files at r1, 20 of 20 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84373#-:-N71ftzM7oaPdX_7kWSA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84373)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/sem/tree/type_check.go` line 1044 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84373#-N6siRT2umqfK3zaLuyM-r2-1044:-N71hlwpCNUfWKxJPRSl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/340fe616de6bdd075e1da64ae72524a19e6da058/pkg/sql/sem/tree/type_check.go#L1044)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

nit: merge this `if` block with the one setting search path above?
</blockquote></details>

Done.

**Comment by mgartner** [comment]
TFTRs! bors r+

**Pull request merged**